### PR TITLE
Add the ability to force push a branch.

### PR DESCRIPTION
**Pull request by dfedde-pivotal**
Add the 'force' option to the put script allowing the user to force
the branch to be pushed to the remote.

### CODE REVIEW

File: test/put.sh [hunk 1]
```diff
@@ -293,6 +293,33 @@ it_can_put_and_set_git_config() {
...
+  cp ~/.gitconfig ~/.gitconfig.orig
+
+  put_uri_with_force $repo1 $src repo | jq -e "
+    .version == {ref: $(echo $ref | jq -R .)}
+  "
+
+  # switch back to master
+  git -C $repo1 checkout master
+
+  test "$(git config --global core.pager)" == 'true'
```
Comment (vito) [inline test/put.sh]: don't need these here - some of the rest might also be copypasta. maybe just base it on the simplest push case?

File: assets/out [hunk 2]
```diff
@@ -79,7 +80,11 @@ tag() {
...
 push_src_and_tags() {
-  git push --tags push-target HEAD:refs/heads/$branch
+  if [ "$force" = "true" ]; then
+    git push --tags push-target HEAD:refs/heads/$branch --force
+  else
+    git push --tags push-target HEAD:refs/heads/$branch
+  fi
 }
 
 push_tags() {
```
Comment (vito) [inline assets/out]: may also be worth updating this for the tag-only push case.

maybe just extract a `$forceflag` var that's either ` --force` or empty, and interpolate it into the arg string for both `push`es?

File: test/put.sh [hunk 3]
```diff
@@ -293,6 +293,33 @@ it_can_put_and_set_git_config() {
...
+  local repo1=$(init_repo)
+
+  local src=$(mktemp -d $TMPDIR/put-src.XXXXXX)
+  local repo2=$src/repo
+  git clone $repo1 $repo2
+
+  local ref=$(make_commit $repo2)
+  make_commit $repo1
+  # cannot push to repo while it's checked out to a branch
+  git -C $repo1 checkout refs/heads/master
```
Comment (vito) [inline test/put.sh]: imo it'd be clearer if this tested the more common case of force-pushing over commits - this is kind of a lesser-known behavior of local directory pushes

File: assets/out (hunk 2 above)
Comment (dfedde-pivotal) [inline assets/out]: Re: updating the `tag-only` push case -- would you ever need to force push if you're only updating tags?

File: test/put.sh (hunk 3 above)
Comment (dfedde-pivotal) [inline test/put.sh]: We thought that is what is happening. We are making a commit on both sides and should lose the commit in repo1. Would checking that the commit made in repo1 is gone in repo2 satisfy this concern?

File: assets/out (hunk 2 above)
Comment (vito) [inline assets/out]: Yeah - you need to force-push if the tag already exists.

File: test/put.sh (hunk 3 above)
Comment (vito) [inline test/put.sh]: Oh my bad - I misread and thought this commented line was setting up the force-push case. I see now that it's doing exactly what I described.

**Comment by dfedde-pivotal** [comment]
We tried to address all of the issues you had. would you take a look at it again?

**Comment by vito** [comment]
looks great, thanks a bunch!

**Pull request merged**